Setup MPU for code relocated to SRAM

mmahadevan108 opened the pull request:
1. Create a static MPU map for code relocated to SRAM
2. Update the python script to provide proper alignment for the SRAM region
3. Setup the static MPU regions before PRE_KERNEL_1 and PRE_KERNEL_2 functions are invoked. This will setup the MPU for SRAM regions in case code relocated to SRAM and invoked from any of these functions.

mmahadevan108 committed:
arch: arm: Create a MPU entry for relocated code

Code relocated using CONFIG_CODE_DATA_RELOCATION_SRAM should
be allowed to execute from SRAM

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
scripts: Set align size for code relocated

Code relocated to SRAM need MPU setup.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

ioannisg commented on kernel/init.c:
```diff
@@ -74,7 +74,6 @@ K_KERNEL_STACK_ARRAY_DEFINE(z_interrupt_stacks, CONFIG_MP_NUM_CPUS,
 
 extern void idle(void *unused1, void *unused2, void *unused3);
 
-
```
@mmahadevan108 can you remove the white space diff from init.c? that is going to remove some of the review requests, as well :)

mmahadevan108 replied:
Thank you for catching this. This was accidentally added.

ioannisg commented on arch/arm/include/aarch32/kernel_arch_func.h:
```diff
@@ -42,6 +43,16 @@ static ALWAYS_INLINE void arch_kernel_init(void)
 	z_arm_fault_init();
 	z_arm_cpu_idle_init();
 	z_arm_clear_faults();
+#if defined(CONFIG_MPU)
```
CONFIG_ARM_MPU here

mmahadevan108 replied:
Thank you, fixed.

ioannisg commented on arch/arm/core/aarch32/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -82,6 +87,14 @@ static const struct z_arm_mpu_partition static_regions[] = {
 		.attr = K_MEM_PARTITION_P_RX_U_RX,
 		},
 #endif /* CONFIG_ARCH_HAS_RAMFUNC_SUPPORT */
+#if defined(CONFIG_CODE_DATA_RELOCATION)
```
This is no different from the RAMFUNC support, at least in the sense of memory attributes. Would it be possible to group these things together?

mmahadevan108 replied:
I don't know of a clean way to do this.

MaureenHelm replied:
This would make sense if the SRAM region was always the relocation destination, but it's not. ITCM is another possibility for some platforms.

ioannisg replied:
understood, thanks

mmahadevan108 replied:
To address this issue I have created a new Kconfig called `CODE_DATA_RELOCATION_SRAM`. The changes made to `arm_code_mpu.c` file are now using` #if defined(CONFIG_CODE_DATA_RELOCATION_SRAM)` which allows the MPU entry to be added only when code is relocated to SRAM.

ioannisg commented on arch/arm/core/aarch32/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -55,6 +55,11 @@ extern K_THREAD_STACK_DEFINE(z_main_stack, CONFIG_MAIN_STACK_SIZE);
 #endif
 
+#if defined(CONFIG_CODE_DATA_RELOCATION)
+extern char __sram_text_start[];
```
Are these symbols defined anywhere?

mmahadevan108 replied:
They get generated at build time and are placed in `build/zephyr/include/generated/linker_relocate.ld`

MaureenHelm replied:
This assumes we always relocate to the SRAM region, which isn't necessarily true. It's possible to relocate to other regions.

ioannisg commented:
I need to have a closer look into this. For now, I could only comment that the refactor you're doing seems to break the support in NXP MPU to start the MPU stuff at PRE KERNEL 2.

Also, for the whole thing to work efficiently, you probably need to have the RAM area for relocation inside the RAM_SECTIONS - is it the case?

mmahadevan108 commented:
> RAM_SECTIONS

Relocation is done using `zephyr_code_relocate(src/*.c SRAM)` function which places the files in SRAM region. `scripts/gen_relocate_app.py` will generate the `linker_relocate.ld` and `code_relocation.c` files at build time.

mmahadevan108 commented:
Rebased

mmahadevan108 commented:
Any comments?

mmahadevan108 commented:
> I need to have a closer look into this. For now, I could only comment that the refactor you're doing seems to break the support in NXP MPU to start the MPU stuff at PRE KERNEL 2.

I have ran userspaces tests on the FRDM_K64F board and the tests pass. The tests I ran were
```
tests/kernel/mem_protect/userspace
samples/userspace/shared_mem
samples/userspace/hello_world_user
samples/userspace/prod_consumer/
```

ioannisg commented:
> > I need to have a closer look into this. For now, I could only comment that the refactor you're doing seems to break the support in NXP MPU to start the MPU stuff at PRE KERNEL 2.
> 
> I have ran userspaces tests on the FRDM_K64F board and the tests pass. The tests I ran were
> 
> ```
> tests/kernel/mem_protect/userspace
> samples/userspace/shared_mem
> samples/userspace/hello_world_user
> samples/userspace/prod_consumer/
> ```

I am not saying that the tests are breaking. I only commented that there was no option any more for POST KERNEL initialization of the module. :)

mmahadevan108 commented:
> > > I need to have a closer look into this. For now, I could only comment that the refactor you're doing seems to break the support in NXP MPU to start the MPU stuff at PRE KERNEL 2.
> > 
> > 
> > I have ran userspaces tests on the FRDM_K64F board and the tests pass. The tests I ran were
> > ```
> > tests/kernel/mem_protect/userspace
> > samples/userspace/shared_mem
> > samples/userspace/hello_world_user
> > samples/userspace/prod_consumer/
> > ```
> 
> I am not saying that the tests are breaking. I only commented that there was no option any more for POST KERNEL initialization of the module. :)

The NXP MPU driver was delaying the init to PRE_KERNEL2 for logging purposes. I think this was meant to help debugging during MPU driver init.

mmahadevan108 commented:
Any questions on this PR?

mmahadevan108 commented:
Any questions on this PR?

ioannisg commented on arch/arm/core/aarch32/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -82,6 +87,14 @@ static const struct z_arm_mpu_partition static_regions[] = {
 		.attr = K_MEM_PARTITION_P_RX_U_RX,
 		},
 #endif /* CONFIG_ARCH_HAS_RAMFUNC_SUPPORT */
+#if defined(CONFIG_CODE_DATA_RELOCATION)
```
The reason that this generally works today is that all the "static" regions are located (forced by the linker, actually) in the beginning of SRAM, before the kernel ram (which includes the application memory). This placement is a requirement, currently, for Cortex-M33, because the static memory regions are actually partitioning the background SRAM, because overlapping regions are not allowed. (btw, you cannot currently call configure_static_regions for Cortex-M33 for a region that is not in RAM, unless i am missing something).

In ARMv7-M this is not a requirement, as the MPU static regions do not need to partition the SRAM, and you can configure static regions basically everywhere.

So I am not sure if this solution is generic enough to be accepted. 

@MaureenHelm  @mmahadevan108 did you try to test this PR on a Cortex-M33 with MPU? Especially if the relocation is not on SRAM. If so, I 'd like to hear how this works.

Besides this, I have no more objections for this change.

mmahadevan108 replied:
I am running this code on a Cortex-M33. The code is relocated to SRAM.

MaureenHelm replied:
We relocate to ITCM instead of SRAM on i.MX RT1064, which is a Cortex-M7. Not an M33. We talked about it a bit in https://github.com/zephyrproject-rtos/zephyr/pull/31132#discussion_r553349992 when I tried to add an MPU region for ITCM, but we then decided it wasn't necessary.

ioannisg replied:
I recall it now, thanks.
In my opinion, we should make sure the solution that is presented here works in a generic way. If not, then we need to restrict it on the ARCHes that support it.

mmahadevan108 replied:
To address this issue I have created a new Kconfig called CODE_DATA_RELOCATION_SRAM. The changes made to arm_code_mpu.c file are now using #if defined(CONFIG_CODE_DATA_RELOCATION_SRAM) which allows the MPU entry to be added only when code is relocated to SRAM.

mmahadevan108 committed:
Kconfig: Add a new config CODE_DATA_RELOCATION_SRAM

1. This will help us identify if the relocation is to
SRAM which is used when setting up the MPU entry
for the SRAM region where code is relocated
2. Move CODE_DATA_RELOCATION configs to ARM specific
folder

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: flash: MCUX: Split the relocate define based on target memory

Use the CODE_DATA_RELOCATION_SRAM config to indentify code relocated
to SRAM so we can setup the MPU for the SRAM region used for code
relocation.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
arch: arm: Setup Static MPU regions earlier in boot flow

Setup the static MPU regions before PRE_KERNEL_1 and
PRE_KERNEL_2 functions are invoked. This will setup
the MPU for SRAM regions in case code relocated to SRAM
is invoked from any of these functions.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

ioannisg commented on Kconfig.zephyr:
```diff
@@ -71,6 +71,17 @@ config CODE_DATA_RELOCATION
 	  files should be specified in the CMakeList.txt file with
 	  a cmake API zephyr_code_relocate().
 
+config CODE_DATA_RELOCATION_SRAM
```
I believe this is only supported in Cortex-M. Maybe you can make the dependency more strict.

Then: for now, could you move this out of Kconfig.zephyr and place the Kconfig definition in the soc/arm/Kconfig instead? Like this we won't be polluting the main zephyr's Kconfig with SoC-specific Kconfig definitions. I suppose that if this is supported in other ARCHes, later, we can move this Kconfig back here. And you could add the CORTEX_M prefix to the name, to make it more clear (that's optional)

MaureenHelm replied:
`CODE_DATA_RELOCATION` is also only supported in Cortex-M (see a few lines up), so I think we should keep these two config definitions together.

ioannisg replied:
OK, that's also fine. Is it possible to move both then, out of Kconfig.zephyr, into Cortex-M arch Kconfig then?

mmahadevan108 replied:
Wouldn't it be better if we kept the 2 configs together in the same place.

mmahadevan108 replied:
Done

mmahadevan108 commented:
Apologies to keep asking about this PR, I would like to get this in before the merge window closes. Also a few other open PR's are dependent on this.

ioannisg commented on arch/arm/core/aarch32/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -82,6 +87,14 @@ static const struct z_arm_mpu_partition static_regions[] = {
+		/* RAM area for relocated text */
+		.start = (uint32_t)&__sram_text_start,
+		.size = (uint32_t)&__sram_text_size,
+		.attr = K_MEM_PARTITION_P_RX_U_RX,
```
This partition has exactly the same attribution as the ramfunc section. I wonder if that one could be re-used. Not against, for now, to accept adding one more region. But i really do not see why we need to have 2 static regions with the same attribution.
